fix: always end sync, even when there is missing data

okdistribute opened the pull request:
I added a temporary hack, based on conversations with @noffle, who said that making hypercore sparse might be more to to chew than we can at the moment... a proper fix can come in v9

1. I backported the `remote-feeds` event from multifeed v5 into multifeed v4.
2. Updated sync progress to include upload as well as download events.
3. Added a heartbeat to the sync -- every 20s, it checks to see if it has had any progress (media or db#upload or db#download) events in the past 20s. If not, it aborts the sync.

okdistribute committed:
test: failing to end on missing data

noffle commented:
Great catch @okdistribute! Things I see we need to do now to fix this:

- [ ] backport https://github.com/kappa-db/multifeed/pull/46 to the version of multifeed we use (pre 5.x.x I think?)
- [ ] pass down `opts.sparse = true` all the way down the stack to hypercore (kappa-osm -> kappa-core -> multifeed -> hypercore)
- [ ] figure out how to know when the multifeed handshake is done & we have all feed refs (I added this to 5.x.x, we'll need to backport)
- [ ] figure out what blocks the remote feeds have
- [ ] modify mapeo-core to issue `core.download()` requests for only those blocks

okdistribute committed:
fix: throw error on missing data

okdistribute committed:
add multifeed dep

okdistribute committed:
fix: make sync timeout heartbeat longer

okdistribute commented:
I added a temporary hack, based on conversations with @noffle, who said that making hypercore sparse might be more to to chew than we can at the moment...

1. I backported the `remote-feeds` event from multifeed into v4.
2. Updated sync progress to include upload as well as download events.
3. Added a heartbeat to the sync -- every 20s, it checks to see if it has had any progress (media or db#upload or db#download) events in the past 20s. If not, it aborts the sync.

hackergrrl commented on lib/db-sync-progress.js:
```diff
@@ -5,17 +5,19 @@ module.exports = function (db, opts) {
   multifeed.ready(function () {
     multifeed.feeds().forEach(onFeed)
-    multifeed.on('feed', onFeed)
+    stream.on('remote-feeds', () => {
```
Could you remove this listener on `eos` below also?

hackergrrl commented on lib/db-sync-progress.js:
```diff
@@ -24,19 +26,26 @@ module.exports = function (db, opts) {
 
   function onFeed (feed) {
-    feeds.push(feed)
+    if (feeds.has(feed)) return
```
Is this working as-is? I'd expect this to be

```suggestion
    if (feeds.has(feed.key.toString('hex'))) return
```

I was going to say use `feed.key` as the map key, but I just checked and it turns out buffers as keys only work if you pass in the *same* buffer reference. Using the hex key is, I think, the safest.

hackergrrl commented on lib/db-sync-progress.js:
```diff
@@ -24,19 +26,26 @@ module.exports = function (db, opts) {
-    progress[feed.key.toString('hex')] = feed.downloaded(0, feed.length)
-    var total = feeds.reduce(function (acc, feed) { return acc + feed.length }, 0)
-    var sofar = feeds.reduce(function (acc, feed) { return acc + feed.downloaded(0, feed.length) }, 0)
+    var all = Array.from(feeds.values())
```
I didn't know about this trick!

hackergrrl commented on lib/db-sync-progress.js:
```diff
@@ -24,19 +26,26 @@ module.exports = function (db, opts) {
+      return acc + feed.length
+    }, 0)
+    var sofar = all.reduce(function (acc, feed) {
+      return acc + feed.stats.totals.downloadedBlocks + feed.stats.totals.uploadedBlocks
```
I *think* this might give misleading results?

My understanding of how `stats.totals` works is that it records every upload and download, across multiple peer sessions. We share the same multifeed and thus the same hypercore references across all syncs, so if someone syncs with on peer they'd maybe see their `uploadedBlocks` go to `100`, but on a second sync (/wo restarting Mapeo), they'll see `uploadedBlocks` **start** at `100`.

In the `v9` branch, I include the totals in the handshake, and count 'upload' and 'download' events. Maybe we just continue to track downloads only for now, but do use the 'upload' events for detection of inactivity?

okdistribute replied:
sounds good

hackergrrl commented on sync.js:
```diff
@@ -448,11 +455,19 @@ class Sync extends events.EventEmitter {
+          // Ideally, we'd open a sparse hypercore instead.
+          heartbeat = setInterval(() => {
+            if (self.state.stale(peer)) {
+              connection.destroy(new Error('missing data'))
```
Maybe..?

```suggestion
              connection.destroy(new Error('timed out due to missing data'))
```

okdistribute committed:
Update sync.js

Co-authored-by: Kira Oakley <noffle@users.noreply.github.com>

okdistribute committed:
Update lib/db-sync-progress.js

Co-authored-by: Kira Oakley <noffle@users.noreply.github.com>

okdistribute committed:
fixes

okdistribute committed:
fix: after feedback from nettle

okdistribute committed:
chore(deps): use multifeed@4.3.0
